// @blueislandx/island-union-sdk/union/platforms/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.youzan = exports.vip = exports.taobao = exports.suning = exports.pinduoduo = exports.kaola = exports.jd = void 0;
var jd_1 = require("./jd");
Object.defineProperty(exports, "jd", { enumerable: true, get: function () { return jd_1.jd; } });
var kaola_1 = require("./kaola");
Object.defineProperty(exports, "kaola", { enumerable: true, get: function () { return kaola_1.kaola; } });
var pinduoduo_1 = require("./pinduoduo");
Object.defineProperty(exports, "pinduoduo", { enumerable: true, get: function () { return pinduoduo_1.pinduoduo; } });
var suning_1 = require("./suning");
Object.defineProperty(exports, "suning", { enumerable: true, get: function () { return suning_1.suning; } });
var taobao_1 = require("./taobao");
Object.defineProperty(exports, "taobao", { enumerable: true, get: function () { return taobao_1.taobao; } });
var vip_1 = require("./vip");
Object.defineProperty(exports, "vip", { enumerable: true, get: function () { return vip_1.vip; } });
var youzan_1 = require("./youzan");
Object.defineProperty(exports, "youzan", { enumerable: true, get: function () { return youzan_1.youzan; } });
